redskyctl/internal/commands/generate/rbac.go

```diff
@@ -90,88 +96,174 @@ func (o *RBACOptions) Complete() {
 }
 
 func (o *RBACOptions) generate() error {
-	// We need to know what namespace the controller is supposed to be in
-	ctrl, err := config.CurrentController(o.Config.Reader())
-	if err != nil {
+	// Read the experiments
+	// TODO For now just pretend like `readExperiment` could return multiple results
+	experimentList := &redskyv1alpha1.ExperimentList{}
+	experimentList.Items = make([]redskyv1alpha1.Experiment, 1)
+	if err := readExperiment(o.Filename, o.In, &experimentList.Items[0]); err != nil {
 		return err
 	}
 
-	// Generate a cluster role
-	clusterRole := &rbacv1.ClusterRole{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: o.Name,
-		},
+	// Determine the binding targets
+	roleRef, subject, namespaces, err := o.bindingTargets(experimentList)
+	if err != nil {
+		return err
 	}
 
-	// The controller uses the default service account for the namespace it is installed in
-	clusterRoleBinding := &rbacv1.ClusterRoleBinding{
-		Subjects: []rbacv1.Subject{
-			{
-				Kind:      "ServiceAccount",
-				Name:      "default",
-				Namespace: ctrl.Namespace,
-			},
-		},
-		RoleRef: rbacv1.RoleRef{
-			APIGroup: "rbac.authorization.k8s.io",
-			Kind:     "ClusterRole",
-		},
+	// Discover the policy rules from the experiments and collapse them
+	var experimentRules []*rbacv1.PolicyRule
+	for i := range experimentList.Items {
+		experimentRules = o.appendRules(experimentRules, &experimentList.Items[i])
 	}
-
-	// Read the experiment
-	// TODO When "readExperiment" returns multiple results, we still just generate a single cluster role
-	experiment := &redskyv1alpha1.Experiment{}
-	if err := readExperiment(o.Filename, o.In, experiment); err != nil {
-		return err
+	rules := make([]rbacv1.PolicyRule, 0, len(experimentRules))
+	for _, r := range experimentRules {
+		rules = mergeRule(rules, r)
+	}
+	if len(rules) == 0 {
+		return nil
 	}
 
-	// Come up with a default name if necessary
-	if clusterRole.Name == "" {
-		if experiment.Name != "" { // TODO Only take the experiment name if it is the only one `&& len(experiments) == 1`
-			clusterRole.Name = "redsky-patching-" + strings.ReplaceAll(strings.ToLower(experiment.Name), " ", "-")
+	// Add up all the objects and print them out
+	rbac := buildRBAC(roleRef, subject, rules, namespaces)
+	return o.Printer.PrintObj(rbac, o.Out)
+}
+
+func (o *RBACOptions) bindingTargets(experimentList *redskyv1alpha1.ExperimentList) (*rbacv1.RoleRef, *rbacv1.Subject, []string, error) {
+	// Create the role reference
+	roleRef := &rbacv1.RoleRef{APIGroup: "rbac.authorization.k8s.io", Kind: "Role", Name: o.Name}
+	if o.ClusterRole {
+		roleRef.Kind = "ClusterRole"
+	}
+	if roleRef.Name == "" {
+		roleRef.Name = "redsky-patching-"
```

Comment:
should we define this as a const?

Reply:
This whole block is a bit iffy. I'm going to move it out to a separate function to clean things up a little.